compiler/noirc_frontend/src/hir_def/types.rs

```diff
@@ -1371,6 +1373,55 @@ impl Type {
         matches!(self.follow_bindings_shallow().as_ref(), Type::Reference(_, _))
     }
 
+    /// Returns `true` if a type is allowed to appear in an assertion message.
+    ///
+    /// This should try filter out types which would cause a panic in `abi_gen::abi_type_from_hir_type`,
+    /// but it has to be more permissive, as we don't have all information yet; some only become apparent
+    /// after monomorphization.
+    pub(crate) fn is_message_compatible(&self, is_monomorphized: bool) -> bool {
+        match self {
+            Type::FieldElement | Type::Integer(_, _) | Type::Bool | Type::String(_) => true,
+
+            Type::Array(_, item) => item.is_message_compatible(is_monomorphized),
+            Type::TypeVariable(binding) => match &*binding.borrow() {
+                TypeBinding::Bound(typ) => typ.is_message_compatible(is_monomorphized),
+                TypeBinding::Unbound(_, kind) => {
+                    matches!(kind, Kind::Integer | Kind::IntegerOrField)
```

Comment:
We should probably return `true` here.

This program:

```noir
fn main() {
    let msg: [_; 3] = std::mem::zeroed();
    assert(false, msg);
    let _ = msg[0] + 1;
}
```

fails with this:

```
error: The type [Field; 3] cannot be used in a message
  ┌─ src/main.nr:3:19
  │
3 │     assert(false, msg);
  │                   ---
```

The problem is that at the type we check the type of `msg`, therre's an unbound type variable. It eventually gets bound to Field, so that should be accepted.